pkg/infradb/bridge.go

```diff
@@ -211,3 +211,54 @@ func (in *LogicalBridge) DeleteBridgePort(bpName, bpMac string) error {
 func (in *LogicalBridge) GetName() string {
 	return in.Name
 }
+
+// setComponentState set the stat of the component
+func (in *LogicalBridge) setComponentState(component common.Component) {
+	lbComponents := in.Status.Components
+	for i, comp := range lbComponents {
+		if comp.Name == component.Name {
+			in.Status.Components[i] = component
+			break
+		}
+	}
+}
+
+// checkForAllSuccess check if all the components are in Success state
+func (in *LogicalBridge) checkForAllSuccess() bool {
+	for _, comp := range in.Status.Components {
+		if comp.CompStatus != common.ComponentStatusSuccess {
+			return false
+		}
+	}
+	return true
+}
+
+// parseMeta parse metadata
+func (in *LogicalBridge) parseMeta(lbMeta *LogicalBridgeMetadata) {
+	if lbMeta != nil {
+		in.Metadata = lbMeta
+	}
+}
+
+// prepareObjectsForReplay prepares an object for replay by setting the unsuccessful components
+// in pending state and returning a list of the components that need to be contacted for the
+// replay of the particular object that called the function.
+func (in *LogicalBridge) prepareObjectsForReplay(componentName string, lbSubs []*eventbus.Subscriber) []*eventbus.Subscriber {
+	// We assume that the list of Components that are returned
+	// from DB is ordered based on the priority as that was the
+	// way that has been stored in the DB in first place.
+	lbComponents := in.Status.Components
+	tempSubs := []*eventbus.Subscriber{}
+	for i, comp := range lbComponents {
+		if comp.Name == componentName || comp.CompStatus != common.ComponentStatusSuccess {
+			in.Status.Components[i] = common.Component{Name: comp.Name, CompStatus: common.ComponentStatusPending, Details: ""}
+			tempSubs = append(tempSubs, lbSubs[i])
+		}
+	}
+	if in.Status.LBOperStatus == LogicalBridgeOperStatusUp {
+		in.Status.LBOperStatus = LogicalBridgeOperStatusDown
+	}
+
+	in.ResourceVersion = generateVersion()
+	return tempSubs
+}
```

Comment:
I have spent some thoughts on this myself in the past. I haven't concluded myself if that will be better or will increase the difficulty of the readability of the code. The way that is now helps us to debug any potential issues better as we have heavy functional deliveries until the end of the year.  I would prefer to leave it as it is for now that we are more focused on the functional deliveries. We can revisit this topic in the future that we will have more time and we can include it in the general restructuring topic. I can open an issue for this to track it if you want

Reply:
Each such duplication increases maintenance efforts and complexity. I do not see anyone is working on it and it annoys me

Reply:
I can fully understand your frustration. We also on our side have limited resources and is inevitable some tasks to have bigger priorities than other. The restructuring we always have in the back of our minds but we just didn't find time or resources to assign it to somebody. I would raise the issue internally to see what I can do about this. 

Also about those functions that you have mentioned above. Maybe they look similar but I will not say that it is duplication because they handle different things which are related to the object that they are dealing with. So from my perspective keeps the code cleaner. Yes maybe they have the same name but they are methods to different structures and their bodies I am not sure if will be made more confusing if we try to fit them in a common function. I can take a look on this and see if I can somehow make something better.

Reply:
Hello @artek-koltun,

I have taken a look into prepareObjectsForReplay function. I really didn't find any sart way to avaoid having the function as a method of the different objects (VRF,LB,BP,SVI) can you please also check and tell me if you have something to suggest ? The way that is now seems ok to me but if you know any better way please let me know.

Reply:
maybe some variation of [template method](https://refactoring.guru/design-patterns/template-method) help here?

Reply:
I checked the template method that you pointed out above. 

This template method wroks better when you have native support for inheritance in the language (e.g. Java). It is a bit tricky and cumbersome to do it in golang and I am not 100% sure that will make the code simpler and more readbale.  Either way if we want to go down that path I believe that this needs to happen properly as a refactoring activity as requires changes in many places which exceed the limits of this feature PR. Also requires proper planning and thinking do be sure how we want to structure the code in order to make it more flexible and also not breaking anything. 

This is an activity that I do not think is suited for this PR and also can break the functionality easily. I think is more safe to do it later as refactoring activity so we have full control on what we are doing and aslo be sure that we do not break anything as we will have more time in our hands to do it properly. I can open an issue to put a hint on the template method that you have pointed out so in the future we know where to look.

Reply:
I checked the template method that you pointed out above. 

This template method works better when you have native support for inheritance in the language (e.g. Java). It is a bit tricky and cumbersome to do it in golang and I am not 100% sure that will make the code simpler and more readable. Either way if we want to go down that path I believe that this needs to happen properly as a refactoring activity as requires changes in many places which exceed the limits of this feature PR. Also requires proper planning and thinking to be sure how we want to structure the code in order to make it more flexible and also not breaking anything. 

This is an activity that I do not think is suited for this PR and also can break the functionality easily. I think is more safe to do it later as refactoring activity so we have full control on what we are doing and aslo be sure that we do not break anything as we will have more time in our hands to do it properly. I can open an issue and put a hint on the template method that you have pointed out so in the future we know where to look. 

Bottom line is that we need to have a proper design first and be sure that this template method works without any problem in golang before we go in and starting doing functional changes. So I think this is something for the future where we will have enough time to think about this and decide what changes we need to do in general in order to make the code better in many places.

Reply:
> This template method works better when you have native support for inheritance in the language (e.g. Java).

Go embedding and interfaces are even more flexible than inheritance in Java and I believe that template method works well for Go as well

I think the commit you provided is only a midway of a rework. `setStatusComponents`, `getStatusComponents`, `isOperationalStatus`, `setOperationalStatus`, `setNewResourceVersion` look very similar in all types, and it means that we could move the impl into the `Domain` struct with related struct members needed to run those methods. And `Domain` will be embedded into vrf/svi etc. (in the language of OOP it is very close to "vrf inherits `Domain`"). Also, it would require to have a single enum for status for all types I think it won't be a big problem since I see in the switches only up, down, toBeDeleted and unspecified. Are those enums used in proto representation?
Basically, templated method is not needed here, we can proceed with bare embedding.

It is true that rework can be broader than the scope of this PR. What I want you to do is to branch off here and create a draft PR where we could start to experiment. The commit here can be deleted.

Reply:
yes the Enums are in the proto definition. I think we could change them and use one generic Enum. I need to think about this

https://github.com/opiproject/opi-api/blob/main/network/evpn-gw/l3_xpu_infra_mgr.proto#L338

I have created an experimentation branch: https://github.com/mardim91/opi-evpn-bridge/tree/experiment-refactoring

we can use this one for the refactoring attempts

Reply:
pls note it and other topics somewhere in the issue, or in that dedicated PR